query/src/main/java/org/infinispan/query/impl/EntityLoader.java

```diff
@@ -23,20 +25,29 @@ public EntityLoader(AdvancedCache<?, ?> cache, KeyTransformationHandler keyTrans
       this.cache = cache;
    }
 
+   private Object decodeKey(EntityInfo entityInfo) {
+      return keyTransformationHandler.stringToKey(entityInfo.getId().toString(), cache.getClassLoader());
+   }
+
    public Object load(EntityInfo entityInfo) {
-      Object cacheKey = keyTransformationHandler.stringToKey(entityInfo.getId().toString(), cache.getClassLoader());
-      return cache.get(cacheKey);
+      return cache.get(decodeKey(entityInfo));
    }
 
-   public List<Object> load(Collection<EntityInfo> entityInfos) {
-      ArrayList<Object> list = new ArrayList<Object>(entityInfos.size());
+   public List<Object> load(List<EntityInfo> entityInfos) {
+      int entitiesSize = entityInfos.size();
+      Set<Object> keys = new LinkedHashSet<>(entitiesSize);
       for (EntityInfo e : entityInfos) {
-         Object entity = load(e);
-         if (entity != null) {
-            list.add(entity);
+         keys.add(decodeKey(e));
+      }
+      Map<?, ?> entries = cache.getAll(keys);
+      List<Object> results = new LinkedList<>();
+      for (Object key : keys) {
```

Comment:
For this case, it would be fine.  I was more thinking of the remote getAll case.  It has to know which value mapped to which key so you can properly populate the getAll map that it returns.  But your case you are just returning the values in order so there is no notion of what mapped to what.

I created https://issues.jboss.org/browse/ISPN-5427 to look into this further.